app/src/main/java/com/techcourse/service/UserService.java

```diff
@@ -24,9 +29,33 @@ public void insert(final User user) {
     }
 
     public void changePassword(final long id, final String newPassword, final String createBy) {
-        final var user = findById(id);
-        user.changePassword(newPassword);
-        userDao.update(user);
-        userHistoryDao.log(new UserHistory(user, createBy));
+        Connection connection = null;
+        try {
+            connection = DataSourceConfig.getTxConnection();
+            connection.setAutoCommit(false);
```

Comment:
커넥션 하나를 어플리케이션 전체 시점에서 재사용하는 것으로 보이는데 이번 커밋 단위에 트랜잭션이 적용되어 수행된 이후 다른 쿼리가 실행되면 어떤 현상이 생기나요?

Reply:
오호 많은것을 배워갑니다!

Connection은 기본적을 thread safe하지 못해서 동시성 문제도 생길 수가 있군요. 그리고 
또 transaction이 끝났으면에도 계속 같은 connection을 사용하면 데이터 무결성에도 큰 문제가 생기는 군요! 

전반적으로 코드 수정 반영하겠습니다. 

혹시 추가로 문제가 되는 사항이 생각나신다면 새양이 생각하는 문제점도 듣고 싶어요

Reply:
정확합니다!

파생되는 다른 문제점들도 있겠지만 가장 큰 동시성 문제를 인지하고 수정하는 부분이 제일 중요하다고 생각됩니다!
수정해주신 코드를 보았을 때 제 기준에선 괜찮아 보입니다!! 😄

Reply:
이전에 `Connection` 하나를 가지고 사용했을 때 `setAutoCommit` 을`false` 로 두었다면 다른 스레드에서는 `AutoCommit` 상태인 줄 알고 `commit` 메서드를 호출하지 않고 있을텐데 그냥 `Statement` 에 쿼리만 추가하고 커밋을 되지 않은 상황이 생기는 것이죠!

이외에도 저도 생각하지 못한 문제가 있을 텐데 이러한 이유들 속에서도 `Connection` 은 풀로 놔두고 사용하는 것도 이해되더라구요!

따라서 수동으로 커밋을 관리할 경우 `setAutoCommit` 이후 `finally` 문에서 다시 `true`로 바꿔줘야할 것 같았어요!
하지만 커넥션 생성 또는 커넥션 풀로 관리하는게 베스트겠죠! 😃

`setAutoCommit` 관련 참고 레퍼런스 였습니다!
https://docs.oracle.com/en/java/javase/21/docs/api/java.sql/java/sql/Connection.html#setAutoCommit(boolean)
> Otherwise, its SQL statements are grouped into transactions that are terminated by a call to either the method commit or the method rollback. By default, new connections are in auto-commit mode.